File: src/engine/run/regionGraph.ts
/**
 * REGION GRAPH GENERATOR — the TypeScript port of ticket 07's ratified prototype.
 *
 * This is the thing the player walks: three sequential biomes of five layers each, generated once
 * at run start from `IRunState.seed` and stored whole in the run save. Nothing here rolls what is
 * *inside* a node — ticket 07 rules that contents are rolled at entry from the node's seed plus its
 * visit count, so that re-entry re-rolls honestly instead of replaying a cached encounter. This
 * module only decides shape and `kind`, both of which are public information the moment the map
 * is drawn (`exploration-map.md`: "types visible, contents hidden").
 *
 * **Every number in `REGION_PARAMS` is a value Henry ruled on 2026-08-21, not a guess.** They come
 * from [ticket 07](../../../docs/wayfinder/steam-release/tickets/07-region-graph.md), which he
 * closed after reacting to rendered graphs from the seeded Python prototype at
 * `docs/wayfinder/steam-release/research/07-region-graph-prototype/regiongraph.py`. Tuning them is
 * a design decision, which is why they are one exported object rather than literals sprinkled
 * through the generator: ticket 10 builds the map screen on this, and whoever tunes the map next
 * should have exactly one place to look.
 *
 * ## Four deliberate departures from the prototype
 *
 * 1. **`layer` is per biome (0–4), not the prototype's global 0–14 counter.** `RegionNodeSchema`
 *    bounds `layer` to 0–4 and carries `biomeIndex` separately, so position is (biome, layer).
 * 2. **No `x`, and no `'entry'` kind.** `x` was dropped because storing a lane index freezes a
 *    layout decision the UI has not made (ticket 06's note on why `layer` replaces raw x/y), and
 *    `NODE_KINDS` has no `'entry'`. Biome 0's layer-0 node is therefore a `wild` that starts at
 *    `visited: 1` — the run begins standing on it, so it does not trigger. Biomes 1 and 2 keep the
 *    prototype's behaviour: their layer-0 node is a `wild` connector you actually fight through.
 * 3. **Edges live on the node and point both ways.** The prototype kept a separate directed edge
 *    list; `IRegionNode.edges` is the only edge storage here, and ticket 07's "walkable in both
 *    directions" is enforced by construction — `link()` writes both halves or neither.
 * 4. **`'market'` is spelled `'marketplace'`**, matching `NodeKind`.
 *
 * One further correction, worth flagging rather than burying: the prototype built its weighted kind
 * pool as `int(weight * 20)` repeats, which truncates 14%/8% down to 2 and 1 entries out of 18 and
 * quietly turns the ruled 60/14/10/8/8 into roughly 67/11/11/6/6. This port uses the ruled
 * percentages as literal weights, so the mix is Henry's number rather than the prototype's
 * rounding error.
 *
 * Engine module: no React, no Redux, no `src/ui` or `src/debug` imports, and no `Math.random` —
 * everything procedural threads through `SeedStream` so a run replays identically from its seed.
 */

import { SeedStream } from '../core/SeedStream';
import type { IRegionNode, NodeKind } from '../runTypes';

// ---------------------------------------------------------------------------------------------
// Ruled parameters (Henry, 2026-08-21, ticket 07)
// ---------------------------------------------------------------------------------------------

export const REGION_PARAMS = {
    /** `exploration-map.md`: a run is three biomes, walked in order. */
    biomesPerRun: 3,
    /** Five layers each: 0 entry, 1–3 middle, 4 exit. `RegionNodeSchema` bounds this to 0–4. */
    layersPerBiome: 5,
    /** Layer indices 1, 2 and 3 — the only layers whose kind is rolled from the mix. */
    middleLayerCount: 3,
    /** "Width: 2–3 nodes per middle layer." Rolled per layer, so biomes are not uniform. */
    minMiddleWidth: 2,
    maxMiddleWidth: 3,
    /**
     * Each node in the previous layer reaches forward to this many nodes in the next one. The
     * prototype's `randint(1, 2)`: 1 keeps routes meaningfully separate, 2 keeps the graph from
     * degenerating into parallel lanes — which is the entire point of "explicitly NOT Spire's
     * three lanes".
     */
    minForwardEdges: 1,
    maxForwardEdges: 2,
    /** "Lateral edges between siblings ~60% of layers." One per layer, between two siblings. */
    lateralEdgeChance: 0.6,
    /**
     * The ruled middle-node mix, as literal percentages. See the header note on why these are not
     * the prototype's quantised pool, and `buildFillerKindPool` for the one place this mix and the
     * market/workshop guarantee contradict each other.
     */
    middleKindWeights: {
        wild: 60,
        event: 14,
        elite: 10,
        marketplace: 8,
        workshop: 8,
    },
    /**
     * "Exactly one market and one workshop guaranteed per biome." These are placed first and the
     * rest of the biome's middle nodes are filled from the weighted pool, so the guarantee is
     * structural rather than a retry loop that might not converge on a narrow biome.
     */
    guaranteedMiddleKinds: ['marketplace', 'workshop'],
    /** "Biome exit = an elite; biome 3's exit is the gym." */
    biomeExitKind: 'elite',
    finalBiomeExitKind: 'gym',
    /** "1 dead-end side node per biome." */
    pocketsPerBiome: 1,
    /**
     * Rolled uniformly from this list, so `wild` lands half the time: a pocket is usually a farming
     * detour and occasionally the alpha (guards a guaranteed blueprint) or the ambush (their 3 vs
     * your 2). Duplicating `wild` rather than writing weights keeps this identical to the
     * prototype Henry actually looked at.
     */
    pocketKinds: ['wild', 'wild', 'alpha', 'ambush'],
} as const;

// ---------------------------------------------------------------------------------------------
// Public shape
// ---------------------------------------------------------------------------------------------

export interface RegionGraph {
    readonly nodes: ReadonlyArray<IRegionNode>;
    /** Biome 0, layer 0 — where the run starts, and what `IRunState.currentNodeId` opens on. */
    readonly entryNodeId: string;
    /** Biome 2, layer 4 — the gauntlet, and the only way a run is won. */
    readonly gymNodeId: string;
}

// ---------------------------------------------------------------------------------------------
// Generation
// ---------------------------------------------------------------------------------------------

/** The generator's working copy: `IRegionNode` is deeply readonly, which is right for consumers. */
interface MutableNode {
    id: string;
    kind: NodeKind;
    biomeIndex: number;
    layer: number;
    pocket: boolean;
    edges: string[];
    visited: number;
}

/**
 * Ticket 07: "the graph is genuinely explorable, not a frontier picker." Writing both halves in
 * one function is the cheapest way to make an asymmetric edge unrepresentable — there is no code
 * path in this module that appends to `edges` directly.
 */
function link(a: MutableNode, b: MutableNode): void {
    if (!a.edges.includes(b.id)) a.edges.push(b.id);
    if (!b.edges.includes(a.id)) b.edges.push(a.id);
}

/**
 * The weighted pool the non-guaranteed middle nodes are drawn from, as a flat array.
 *
 * **The one place ticket 07's ruled parameters contradict themselves, resolved here in the open.**
 * The mix lists marketplace at 8% and workshop at 8%, *and* the same row guarantees "exactly one
 * market and one workshop per biome". Both cannot hold: a biome has 6–9 middle nodes, so if the
 * filler pool can also roll a marketplace, biomes with two of them are common (the prototype has
 * this bug — its pool includes both, and it produces double markets regularly).
 *
 * The guarantee wins, because it is the half with a gameplay argument behind it: a biome with two
 * workshops and a biome with none are both worse than a biome with one, and ticket 07's own test
 * list names "exactly one market + one workshop per biome" as a check the port must pass. So the
 * filler pool drops the two guaranteed kinds and renormalises over what is left — wild, event and
 * elite keep their 60 : 14 : 10 ratio to each other.
 *
 * What that costs, stated rather than hidden: because the guarantee places 2 of an average 7.5
 * middle nodes, the *realised* per-biome mix runs about 13% marketplace and 13% workshop against
 * the ruled 8%, and correspondingly fewer wilds. The ruled percentages describe the pool; the
 * guarantee overrides them at these widths. Widening biomes is the only thing that would close the
 * gap, and that is Henry's call, not this module's — hence `middleKindWeights` keeps the ruled
 * numbers verbatim rather than being quietly rewritten to what the generator actually produces.
 */
function buildFillerKindPool(): NodeKind[] {
    const guaranteed = REGION_PARAMS.guaranteedMiddleKinds as ReadonlyArray<NodeKind>;
    const pool: NodeKind[] = [];
    const weights = Object.entries(REGION_PARAMS.middleKindWeights) as ReadonlyArray<[NodeKind, number]>;
    for (const [kind, weight] of weights) {
        if (guaranteed.includes(kind)) continue;
        for (let i = 0; i < weight; i += 1) pool.push(kind);
    }
    return pool;
}

function pick<T>(stream: SeedStream, items: ReadonlyArray<T>): T {
    return items[stream.nextInt(0, items.length - 1)];
}

/**
 * Build the whole region for a run.
 *
 * Deterministic in `seed` alone: the same seed always produces the same graph, node ids included,
 * which is what lets ticket 23 resume a mid-run app close by storing one seed string plus the node
 * states rather than the pre-rolled world.
 */
export function generateRegionGraph(seed: string): RegionGraph {
    // Fork rather than consuming the run seed directly. Other subsystems (encounter generation,
    // shop stock) are seeded from the same `IRunState.seed`, and a label keeps two of them from
    // drawing the identical number sequence.
    const stream = new SeedStream(new SeedStream(seed).fork('region-graph'));

    const nodes: MutableNode[] = [];
    const pool = buildFillerKindPool();
    /** Per (biome, layer) counter, so ids are readable positions and still unique with pockets. */
    const indexInLayer = new Map<string, number>();

    const add = (biomeIndex: number, layer: number, kind: NodeKind, pocket: boolean): MutableNode => {
        const key = `${biomeIndex}:${layer}`;
        const index = indexInLayer.get(key) ?? 0;
        indexInLayer.set(key, index + 1);
        const node: MutableNode = {
            id: `b${biomeIndex}l${layer}n${index}`,
            kind,
            biomeIndex,
            layer,
            pocket,
            edges: [],
            visited: 0,
        };
        nodes.push(node);
        return node;
    };

    let entryNodeId = '';
    let gymNodeId = '';
    /** The frontier the next layer attaches to. Carries across the biome seam, one exit deep. */
    let prevLayer: MutableNode[] = [];

    for (let biomeIndex = 0; biomeIndex < REGION_PARAMS.biomesPerRun; biomeIndex += 1) {
        const isFinalBiome = biomeIndex === REGION_PARAMS.biomesPerRun - 1;

        // --- Layer 0: the entry ---------------------------------------------------------------
        // `NODE_KINDS` has no `'entry'`, so this is a `wild`. Biome 0's is the one node the player
        // starts standing on: it is marked `visited: 1` so the entry-trigger rule ("entering a node
        // triggers it again, always") does not fire a fight before the run has begun. Biomes 1 and
        // 2 open with a wild you genuinely fight — that connector is part of the fight envelope.
        const entry = add(biomeIndex, 0, 'wild', false);
        if (biomeIndex === 0) {
            entry.visited = 1;
            entryNodeId = entry.id;
        }
        for (const previous of prevLayer) link(previous, entry);
        prevLayer = [entry];

        // --- Layers 1–3: the middle -------------------------------------------------------------
        const middles: MutableNode[] = [];
        for (let layer = 1; layer <= REGION_PARAMS.middleLayerCount; layer += 1) {
            const width = stream.nextInt(REGION_PARAMS.minMiddleWidth, REGION_PARAMS.maxMiddleWidth);
            const layerNodes: MutableNode[] = [];
            // Kind is a placeholder until the whole biome's middle is known — the market/workshop
            // guarantee is a per-biome property, so it cannot be decided one layer at a time.
            for (let i = 0; i < width; i += 1) layerNodes.push(add(biomeIndex, layer, 'wild', false));

            for (const previous of prevLayer) {
                const reach = Math.min(
                    layerNodes.length,
                    stream.nextInt(REGION_PARAMS.minForwardEdges, REGION_PARAMS.maxForwardEdges),
                );
                for (const target of stream.shuffle(layerNodes).slice(0, reach)) link(previous, target);
            }

            // A node nobody reached forward to is unreachable, and an unreachable node is a piece of
            // map the player can see and never touch. Adopt it rather than regenerating the layer.
            for (const target of layerNodes) {
                if (target.edges.length === 0) link(pick(stream, prevLayer), target);
            }

            // Lateral edges are what make this a graph rather than a braid: they let the player
            // switch route mid-layer, which is how a marketplace two lanes over becomes reachable.
            if (layerNodes.length >= 2 && stream.next() < REGION_PARAMS.lateralEdgeChance) {
                const [a, b] = stream.shuffle(layerNodes);
                link(a, b);
            }

            middles.push(...layerNodes);
            prevLayer = layerNodes;
        }

        // --- Kinds for the middle ---------------------------------------------------------------
        // The guarantees go in first and the remainder is drawn from the weighted pool, then the
        // whole list is shuffled onto the nodes. There are at least 3 x 2 = 6 middle nodes per
        // biome, so the two guaranteed kinds always fit.
        const kinds: NodeKind[] = [...REGION_PARAMS.guaranteedMiddleKinds];
        while (kinds.length < middles.length) kinds.push(pick(stream, pool));
        const assigned = stream.shuffle(kinds);
        middles.forEach((node, i) => { node.kind = assigned[i]; });

        // --- The pocket ---------------------------------------------------------------------------
        // A dead end hanging off a middle node, sharing its host's layer because it is beside the
        // route rather than along it. Its single edge is what makes it a decision: everything you
        // spend getting there you spend again coming back.
        for (let i = 0; i < REGION_PARAMS.pocketsPerBiome; i += 1) {
            const host = pick(stream, middles);
            const pocket = add(biomeIndex, host.layer, pick(stream, REGION_PARAMS.pocketKinds), true);
            link(host, pocket);
        }

        // --- Layer 4: the exit --------------------------------------------------------------------
        // Every route through the biome converges here, which is why the elite is unavoidable and
        // why the gym is the only way the run ends in a victory.
        const exit = add(
            biomeIndex,
            REGION_PARAMS.layersPerBiome - 1,
            isFinalBiome ? REGION_PARAMS.finalBiomeExitKind : REGION_PARAMS.biomeExitKind,
            false,
        );
        for (const previous of prevLayer) link(previous, exit);
        if (isFinalBiome) gymNodeId = exit.id;
        prevLayer = [exit];
    }

    return {
        nodes: nodes.map((node): IRegionNode => ({ ...node, edges: [...node.edges] })),
        entryNodeId,
        gymNodeId,
    };
}

File: docs/wayfinder/steam-release/research/07-region-graph-prototype/regiongraph.py
"""Ticket 07 prototype: seeded region-graph generator + SVG dump. Parameters are the knobs Henry rules."""
import random, json, sys

PARAMS = dict(
    biomes=["Fire", "Water", "Nature"],
    layers_per_biome=5,          # entry, mid, mid, exit (exit of biome 3 = the gym)
    width=3,                     # max nodes per middle layer
    pocket_per_biome=1,          # dead-end side nodes (farm / alpha), return to where you came from
    mix=dict(wild=0.60, elite=0.10, market=0.08, event=0.14, workshop=0.08),
    see_ahead=2,                 # layers of node TYPES visible ahead of the player's layer
    workshop_guarantee=True,     # exactly one workshop per biome, placed in a middle layer
    market_guarantee=True,       # exactly one market per biome
)

ICON = dict(wild="W", elite="E", market="$", event="?", workshop="K", gym="G", entry="S", alpha="A", ambush="X")

def gen(seed, P=PARAMS):
    rng = random.Random(seed)
    nodes, edges = [], []
    nid = 0
    def add(biome, layer, kind, x):
        nonlocal nid
        n = dict(id=nid, biome=biome, layer=layer, kind=kind, x=x); nodes.append(n); nid += 1; return n
    prev_layer = None
    glayer = 0
    for bi, biome in enumerate(P["biomes"]):
        # layer 0 of biome: entry (biome 0) or single connector node
        entry = add(biome, glayer, "entry" if bi == 0 else "wild", 1)
        if prev_layer:
            for p in prev_layer: edges.append((p["id"], entry["id"]))
        prev_layer = [entry]; glayer += 1
        mids = []
        for li in range(P["layers_per_biome"] - 2):
            w = rng.randint(2, P["width"])
            layer = [add(biome, glayer, "wild", x) for x in range(w)]
            # connect: each prev node -> 1-2 next nodes, each next node has >=1 incoming
            for p in prev_layer:
                for t in rng.sample(layer, min(len(layer), rng.randint(1, 2))): edges.append((p["id"], t["id"]))
            for t in layer:
                if not any(e[1] == t["id"] for e in edges): edges.append((rng.choice(prev_layer)["id"], t["id"]))
            # lateral edge (explorable, not lanes)
            if len(layer) >= 2 and rng.random() < 0.6:
                a, b = rng.sample(layer, 2); edges.append((a["id"], b["id"]))
            mids += layer; prev_layer = layer; glayer += 1
        # assign kinds to middle nodes
        kinds = []
        if P["workshop_guarantee"]: kinds.append("workshop")
        if P["market_guarantee"]: kinds.append("market")
        pool = [k for k in P["mix"] for _ in range(int(P["mix"][k] * 20))]
        while len(kinds) < len(mids): kinds.append(rng.choice(pool))
        rng.shuffle(kinds)
        for n, k in zip(mids, kinds): n["kind"] = k
        # pockets: dead-end farm nodes hanging off a middle node
        for _ in range(P["pocket_per_biome"]):
            host = rng.choice(mids)
            pk = add(biome, host["layer"], rng.choice(["wild", "wild", "alpha", "ambush"]), -1)
            pk["pocket"] = True; edges.append((host["id"], pk["id"])); edges.append((pk["id"], host["id"]))
        # exit: biome boss-gate node (wild) or the gym
        ex = add(biome, glayer, "gym" if bi == len(P["biomes"]) - 1 else "elite", 1)
        for p in prev_layer: edges.append((p["id"], ex["id"]))
        prev_layer = [ex]; glayer += 1
    return dict(seed=seed, params=P, nodes=nodes, edges=edges)

FIGHT = {"wild", "elite", "alpha", "ambush"}
def fight_envelope(g):
    """min and max fights on any entry->gym path (pockets excluded from min, included once each in max)."""
    from functools import lru_cache
    out = {}
    for a, b in g["edges"]: out.setdefault(a, []).append(b)
    gym = [n for n in g["nodes"] if n["kind"] == "gym"][0]["id"]
    byid = {n["id"]: n for n in g["nodes"]}
    @lru_cache(None)
    def mn(i):
        c = 1 if byid[i]["kind"] in FIGHT else 0
        if i == gym: return c
        nxt = [j for j in out.get(i, []) if not byid[j].get("pocket") and byid[j]["layer"] > byid[i]["layer"]]
        return c + min(mn(j) for j in nxt)
    @lru_cache(None)
    def mx(i):
        c = 1 if byid[i]["kind"] in FIGHT else 0
        if i == gym: return c
        nxt = [j for j in out.get(i, []) if not byid[j].get("pocket") and byid[j]["layer"] > byid[i]["layer"]]
        return c + max(mx(j) for j in nxt)
    pockets = sum(1 for n in g["nodes"] if n.get("pocket") and n["kind"] in FIGHT)
    return mn(0), mx(0) + pockets

COLOR = dict(Fire="#c0392b", Water="#2980b9", Nature="#27ae60")
def svg(g, player_layer=0):
    P = g["params"]; W, H = 1100, 480; ly = max(n["layer"] for n in g["nodes"]) + 1
    sx = W / (ly + 1); sy = 90
    pos = {n["id"]: (sx * (n["layer"] + 1), 140 + sy * n["x"]) for n in g["nodes"]}
    o = [f'<svg xmlns="http://www.w3.org/2000/svg" width="{W}" height="{H}" style="background:#111;font-family:monospace">']
    for a, b in g["edges"]:
        (x1, y1), (x2, y2) = pos[a], pos[b]
        o.append(f'<line x1="{x1}" y1="{y1}" x2="{x2}" y2="{y2}" stroke="#555" stroke-width="2"/>')
    for n in g["nodes"]:
        x, y = pos[n["id"]]; vis = n["layer"] <= player_layer + P["see_ahead"]
        fill = COLOR.get(n["biome"], "#888") if vis else "#333"
        lab = ICON[n["kind"]] if vis else "·"
        r = 22 if n["kind"] in ("gym", "elite", "alpha") else 17
        o.append(f'<circle cx="{x}" cy="{y}" r="{r}" fill="{fill}" stroke="{"#fff" if n.get("pocket") else "#000"}" stroke-width="2"/>')
        o.append(f'<text x="{x}" y="{y+5}" text-anchor="middle" fill="#fff" font-size="15" font-weight="bold">{lab}</text>')
    mn, mx = fight_envelope(g)
    o.append(f'<text x="10" y="{H-12}" fill="#ccc" font-size="13">seed {g["seed"]} · fights to gym: min {mn} / max {mx} (+3 gauntlet) · W wild, E elite, $ market, ? event, K workshop, A alpha, X ambush (white ring = pocket/dead-end), G gym · grey = beyond visibility ({P["see_ahead"]} layers)</text>')
    o.append("</svg>"); return "\n".join(o)

if __name__ == "__main__":
    seeds = [int(s) for s in sys.argv[1:]] or [1, 2, 3]
    for s in seeds:
        g = gen(s); open(f"/root/sr/proto07/graph_{s}.svg", "w").write(svg(g))
        print(s, "nodes", len(g["nodes"]), "fights", fight_envelope(g))
